src/v/security/license.cc

```diff
@@ -0,0 +1,302 @@
+    static ss::sstring public_key_header = "-----BEGIN PUBLIC KEY-----";
+    static ss::sstring public_key_footer = "-----END PUBLIC KEY-----";
+
+    size_t pos1{ss::sstring::npos}, pos2{ss::sstring::npos};
+    pos1 = pem_key.find(public_key_header);
+    if (pos1 == ss::sstring::npos) {
+        throw std::runtime_error(
+          "Embedded public key error: PEM header not found");
+    }
+
+    pos2 = pem_key.find(public_key_footer, pos1 + 1);
+    if (pos2 == ss::sstring::npos) {
+        throw std::runtime_error(
+          "Embedded public key error: PEM footer not found");
+    }
+
+    // Start position and length
+    pos1 = pos1 + public_key_header.length();
+    pos2 = pos2 - pos1;
+    return pem_key.substr(pos1, pos2);
+}
+
+static CryptoPP::ByteQueue convert_pem_to_ber(const ss::sstring& pem_key) {
+    const ss::sstring keystr = parse_pem_contents(pem_key);
+    CryptoPP::StringSource ss{keystr.c_str(), true};
+
+    // Base64 decode, place in a ByteQueue
+    CryptoPP::ByteQueue queue;
+    CryptoPP::Base64Decoder decoder;
+    decoder.Attach(new CryptoPP::Redirector(queue));
+    ss.TransferTo(decoder);
+    decoder.MessageEnd();
+    return queue;
+}
+
+static const CryptoPP::RSA::PublicKey public_key = []() {
+    static const ss::sstring public_key_material
+      = "-----BEGIN PUBLIC KEY-----\n"
+        "MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAt0Y2jGOLI70xkF4rmpNM\n"
+        "hBqU3cUrwYCREgjT9TT77KusvhPVc16cdK83bpaGQy+Or1WyZpN+TCxT2vlaZet6\n"
+        "RDo+55jRk7epazAHx9s+DLd6IzhSXakf6Sxh5JRK7Zn/75C1hYJMspcJ75EhLv4H\n"
+        "qXj12dkyivcLAecGhWdIGK95J0P7f4EQQGwGL3rilCSlfkVVmE4qaPUaLqULKelq\n"
+        "7T2d+AklR+KwgtHINyKDPJ9+cCAMoEOrRBDPjcQ79k0yvP3BdHV394F+2Vt/AYOL\n"
+        "dcVQBm3tqIySLGFtiJp+RIa+nJhMrd+G4sqwm4FhsmG35Fbr0XQJY0sM6MaFJcDH\n"
+        "swIDAQAB\n"
+        "-----END PUBLIC KEY-----\n";
+    auto queue = convert_pem_to_ber(public_key_material);
+    CryptoPP::RSA::PublicKey public_key;
+    public_key.BERDecode(queue);
+    return public_key;
+}();
+
+/// The redpanda license is comprised of 2 sections seperated by a delimiter.
+/// The first section is the data section (base64 encoded), the second being the
+/// signature, which is a PCKS1.5 sigature of the contents of the data section.
+static bool
+verify_license(const ss::sstring& data, const ss::sstring& signature) {
+    CryptoPP::RSASS<CryptoPP::PKCS1v15, CryptoPP::SHA256>::Verifier verifier(
+      public_key);
+    return verifier.VerifyMessage(
+      reinterpret_cast<const CryptoPP::byte*>(data.c_str()), // NOLINT
+      data.length(),
+      reinterpret_cast<const CryptoPP::byte*>(signature.data()), // NOLINT
+      signature.size());
+}
+
+} // namespace crypto
+
+ss::sstring license_type_to_string(license_type type) {
+    switch (type) {
+    case license_type::free_trial:
+        return "free_trial";
+    case license_type::enterprise:
+        return "enterprise";
+    default:
+        __builtin_unreachable();
```

Comment:
are all the cases of the enum covered here and you intended to move `__builtin_unreachable();` outside the case statement and remove the default case in order to make gcc happy? otherwise if you really want to crash-stop on this as a logic error then use a vassert otherwise it will pop up as a generic segfault.

Reply:
ah i didn't know `gcc` could examine for all switch cases, will move to outside